include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -186,297 +186,242 @@ class __kernel_name_base
     {
 #if 0 //_ONEDPL_KERNEL_BUNDLE_PRESENT
         auto __kernel_bundle = sycl::get_kernel_bundle<sycl::bundle_state::executable>(__exec.queue().get_context());
-        return __kernel_bundle.get_kernel(sycl::get_kernel_id<_DerivedKernelName>());
+        return __kernel_bundle.get_kernel(sycl::get_kernel_id<_DerivedKernelName...>());
 #else
         sycl::program __program(__exec.queue().get_context());
 
-        __program.build_with_kernel_type<_DerivedKernelName>();
-        return __program.get_kernel<_DerivedKernelName>();
+        __program.build_with_kernel_type<_DerivedKernelName...>();
+        return __program.get_kernel<_DerivedKernelName...>();
 #endif
     }
 };
 
-template <typename... _Name>
-class __parallel_for_kernel : public __kernel_name_base<__parallel_for_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_reduce_kernel : public __kernel_name_base<__parallel_reduce_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_scan_local_kernel : public __kernel_name_base<__parallel_scan_local_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_scan_global_kernel : public __kernel_name_base<__parallel_scan_global_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_scan_propagate_kernel : public __kernel_name_base<__parallel_scan_propagate_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_find_or_kernel : public __kernel_name_base<__parallel_find_or_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_merge_kernel : public __kernel_name_base<__parallel_merge_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_sort_leaf_kernel : public __kernel_name_base<__parallel_sort_leaf_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_sort_global_kernel : public __kernel_name_base<__parallel_sort_global_kernel<_Name...>>
-{
-};
-template <typename... _Name>
-class __parallel_sort_copy_back_kernel : public __kernel_name_base<__parallel_sort_copy_back_kernel<_Name...>>
-{
-};
 
-//------------------------------------------------------------------------
-// parallel_for - async pattern
-//------------------------------------------------------------------------
 
-//General version of parallel_for, one additional parameter - __count of iterations of loop __cgh.parallel_for,
-//for some algorithms happens that size of processing range is n, but amount of iterations is n/2.
-template <typename _ExecutionPolicy, typename _Fp, typename _Index, typename... _Ranges>
-oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, __future<void>>
-__parallel_for(_ExecutionPolicy&& __exec, _Fp __brick, _Index __count, _Ranges&&... __rngs)
+template <typename _KernelName>
+struct __parallel_for_invoker;
+
+template <typename... _Name>
+struct __parallel_for_invoker<__internal::__optional_kernel_name<_Name...>>
```

Comment:
BTW, upon time I suggested changing "the entry points" of SYCL backend from global functions to a one class "API" and methods. It was because we need at least the second variadic pack  - "dependencies..." for async API.
See, https://github.com/oneapi-src/oneDPL/blob/dev/mdvorski/async_for_each_proto/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h#L239

So, now, we already have tow variadic packs - ranges and names. It seems, In the future , we have to add the third - the "dependencies". => it looks we will need to introduce the third "imbedded" API level like this: "__api<dependencies...>::pattern_invoker<names...>::operator()<...., ranges...>

Reply:
As we discussed it should not be a problem. We have the third parameter pack in specialization only, for primary template it's still one type